fuzzers/run_fuzzer.py

```diff
@@ -0,0 +1,528 @@
+
+     -n == --dry-run
+
+    >>> run_submake('n')
+    False
+    >>> run_submake('n --blah')
+    False
+    >>> run_submake('--blah n')
+    False
+    >>> run_submake('--blah')
+    True
+    >>> run_submake('--random')
+    True
+
+     -q == --question
+
+    >>> run_submake('q')
+    False
+    >>> run_submake('q --blah')
+    False
+    >>> run_submake('--blah q')
+    False
+    >>> run_submake('--blah')
+    True
+    >>> run_submake('--random')
+    True
+
+      Both --dry-run and --question
+
+    >>> run_submake('qn')
+    False
+    >>> run_submake('nq')
+    False
+    >>> run_submake('--quiant')
+    True
+    """
+    r = re.search(r'(?:^|\s)[^-]*(n|q)[^\s]*(\s|$)', make_flags)
+    if not r:
+        return True
+    return not bool(r.groups()[0])
+
+
+def main(argv):
+    fuzzers = os.path.abspath(os.path.dirname(__file__))
+
+    assert len(argv) == 2, argv
+    fuzzer = argv[1]
+
+    make_cmd = os.environ.get('MAKE', 'make')
+    make_flags = os.environ.get('MAKEFLAGS', '')
+
+    # Should run things?
+    if not run_submake(make_flags):
+        return 0
+
+    time_start = datetime.utcnow()
+
+    fuzzer_length = max(len(f) for f in os.listdir(fuzzers))
+
+    logger = Logger(fuzzer, time_start, fuzzer_length)
+
+    def log(msg, *a, **k):
+        logger.log(msg, a, k, flush=True)
+
+    # Need a signal handler on SIGCHLD otherwise get_resource doesn't return
+    # anything.
+    signal.signal(signal.SIGCHLD, lambda sig, frame: None)
+
+    fuzzer_dir = os.path.join(fuzzers, fuzzer)
+    assert os.path.exists(fuzzer_dir), fuzzer_dir
+    # FIXME: The fuzzer output directory should be different from the top level
+    # fuzzer directory.
+    fuzzer_out = fuzzer_dir
+    if not os.path.exists(fuzzer_out):
+        os.makedirs(fuzzer_out)
+    assert os.path.exists(fuzzer_out)
+
+    fuzzer_runok = os.path.join(fuzzer_out, "run.ok")
+
+    if os.path.exists(fuzzer_runok):
+        last_modified = datetime.fromtimestamp(os.stat(fuzzer_runok).st_mtime)
+
+        log(
+            "Skipping as run.ok exists (updated @ {})",
+            last_modified.isoformat())
+    else:
+        log("Starting @ {}", time_start.isoformat())
+
+    log_suffix = ".{}.log".format(time_start.isoformat())
+    fuzzer_stdout = os.path.join(fuzzer_out, "stdout" + log_suffix)
+    fuzzer_stderr = os.path.join(fuzzer_out, "stderr" + log_suffix)
+
+    # Make sure the stdout file exists
+    stdout_fd = open(fuzzer_stdout, "w")
+    stdout_fd.write("Build started {}\n".format(time_start.isoformat()))
+    stdout_fd.write("-" * 75)
+    stdout_fd.write("\n")
+    stdout_fd.flush()
+    os.fsync(stdout_fd)
+    stdout_fd.close()
+
+    stdout_fd = open(fuzzer_stdout, "w+")
+    stdout_fd.seek(0, os.SEEK_END)
+
+    # Make sure the stderr file exists
+    stderr_fd = open(fuzzer_stderr, "w")
+    stderr_fd.write("Build started {}\n".format(time_start.isoformat()))
+    stderr_fd.write("-" * 75)
+    stderr_fd.write("\n")
```

Comment:
Don't write anything to the stderr file.  Eventually we should consider check that this file is empty?

Reply:
The goal with the header was to make sure you have the matching stderr / stdout files.

Reply:
I reworked this code to make that much clear.